File: editor/src/engine/strobeEffect.ts
import { Effect_StrobeEffect } from '@dmx-controller/proto/effect_pb';

import { RenderContext } from './context';
import { applyState } from './effect';

export function strobeEffect(
  context: RenderContext,
  effect: Effect_StrobeEffect,
  frame: number,
): void {
  if (effect.stateA == null || effect.stateB == null) {
    throw new Error('Tried to render strobe effect without state!');
  }
  if (frame % (effect.stateAFames + effect.stateBFames) < effect.stateAFames) {
    applyState(effect.stateA, context);
  } else {
    applyState(effect.stateB, context);
  }
}

File: src-engine/src/render/strobe_effect.rs
use crate::{
    proto::{effect::StrobeEffect, ColorPalette, OutputTarget, Project},
    render::{render_target::RenderTarget, util::apply_state},
};

pub fn apply_strobe_effect<T: RenderTarget<T>>(
    project: &Project,
    render_target: &mut T,
    output_target: &OutputTarget,
    frame: &u32,
    strobe_effect: &StrobeEffect,
    color_palette: &ColorPalette,
) {
    if frame % (strobe_effect.state_a_fames + strobe_effect.state_b_fames)
        < strobe_effect.state_a_fames
    {
        apply_state(
            project,
            render_target,
            output_target,
            strobe_effect.state_a.as_ref().unwrap(),
            color_palette,
        );
    } else {
        apply_state(
            project,
            render_target,
            output_target,
            strobe_effect.state_b.as_ref().unwrap(),
            color_palette,
        );
    }
}
